streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -1210,32 +1205,29 @@ private String decorateTopic(final String topic) {
         return applicationId + "-" + topic;
     }
 
+    void initializeSubscription() {
+        if (usesPatternSubscription()) {
+            log.debug("Found pattern subscribed source topics, initializing consumer's subscription pattern.");
+            final List<String> allSourceTopics = maybeDecorateInternalSourceTopics(sourceTopicNames);
+            Collections.sort(allSourceTopics);
+            sourceTopicPattern = buildPattern(allSourceTopics, nodeToSourcePatterns.values());
+        } else {
+            log.debug("No source topics using pattern subscription found, initializing consumer's subscription collection.");
+            sourceTopicCollection = maybeDecorateInternalSourceTopics(sourceTopicNames);
+            Collections.sort(sourceTopicCollection);
+        }
+    }
+
     boolean usesPatternSubscription() {
         return !nodeToSourcePatterns.isEmpty();
     }
 
     synchronized Collection<String> sourceTopicCollection() {
-        log.debug("No source topics using pattern subscription found, using regular subscription for the main consumer.");
-
-        if (topicCollection == null) {
-            topicCollection = maybeDecorateInternalSourceTopics(sourceTopicNames);
-            Collections.sort(topicCollection);
-        }
-
-        return topicCollection;
+        return sourceTopicCollection;
```

Comment:
Hmm.. inside StreamThread do we guarantee we always call `initializeSubscription` before calling these two functions?

Reply:
If not we should call `initializeSubscription` inside these two as well.

Reply:
I put this initialization inside `build()`, which gets called in the `KafkaStreams` constructor before any `StreamThreads` get created, so the `InternalTopologyBuilder` should always have an initialized subscription by the time it gets passed to a thread.